cmd/govim/gopls_client.go

```diff
@@ -151,9 +151,36 @@ func (g *govimplugin) Configuration(ctxt context.Context, params *protocol.Param
 	return res, nil
 }
 
-func (g *govimplugin) ApplyEdit(context.Context, *protocol.ApplyWorkspaceEditParams) (*protocol.ApplyWorkspaceEditResponse, error) {
+func (g *govimplugin) ApplyEdit(ctxt context.Context, params *protocol.ApplyWorkspaceEditParams) (*protocol.ApplyWorkspaceEditResponse, error) {
 	defer absorbShutdownErr()
-	panic("ApplyEdit not implemented yet")
+	g.logGoplsClientf("ApplyEdit: %v", pretty.Sprint(params))
+
+	var err error
+	var res *protocol.ApplyWorkspaceEditResponse
+	g.applyEditsLock.Lock()
+	if g.applyEditsCh == nil {
+		// ApplyEdit wasn't send by another blocking call so it's fine to just schedule the edits here
+		done := make(chan struct{})
```

Comment:
Where `g.applyEditsCh == nil`, then we don't need to block here. Simply `g.applyEditsLock.Unlock()` then schedule (there is no reason to hold the lock whilst scheduling). We definitely do not need to block until the edit itself is applied.

Reply:
Ah, right. I was thinking about the case where a user invoked `GOVIMFillStruct` the moment before we schedule edits in this branch. And missed the fact that it won't deadlock since it would be an entirely unrelated call.